docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,338 @@
+
+3. Wait for asynchronous resource creation to populate late initialized fields.
+
+    > For some services/resources the late initialized fields do not appear immediately.  This happens when there is asynchronous processing involved in resource creation. The proposed solution should allow waiting and retry until the late initialized fields appear in the mentioned API output.
+
+4. Allow the AWS service teams to provide custom code using hooks for handing of late initialized fields.
+
+## Proposed Solution
+In this solution, There will be updates in 
+1) **ACK Code Generator** to capture the late initialized fields from AWS operation output and use them to late initialize k8s custom resource object. Code generator changes will support a) configuration for service teams to tell ACK code generator which fields will be late initialized and b) configuration to provide custom code for late initialization.
+2) **ACK Runtime** to a) update the custom resource stored in etcd with late initalized optional spec fields And b) handle the late initalized fields for resources that have asynchronous creation.
+    > For the resources with Async creation, the late initalized fields are not readily available after create operation. Therefore the reconciler loop needs to be repeated for these resources until the late initalized fields are available.
+
+### ACK Code Generator Changes
+1. **Provide a new "ServerSideDefaultsConfig" struct for AWS service teams to provide which spec fields(spec.A) or spec field members(spec.A.X) will be late initialized.**
+
+    Only the fields/members mentioned as part of 'ServerSideDefaultsConfig' will be considered for late initialization. (The reason is explained inside step 3 "NOTE" )
+
+    `ServerSideDefaultsConfig` struct will be the member of `ResourceConfig` and it will look like following:
+
+    ```
+    type ServerSideDefaultsConfig struct { 
+    
+    // valid values: 'readOne, readMany, create, update'
+    DefaultSourceMethod *string `json:"default_source_method"`
+    
+    DefaultedSpecFields []DefaultedSpecFieldConfig `json:"defaulted_spec_fields"`
+    }
+
+
+    type DefaultedSpecFieldConfig struct {
+    
+    // valid values: 'readOne, readMany, create, update'
+    // If this is not mentioned, then default_source_method is used
+    SourceMethod *string `json:"source_method,omitempty"`
+    
+    // Path to the field relative to Spec, which will be late initialized
+    // Example for primitives: "a" to replace "desired.ko.Spec.a" with "<source_method>Latest.ko.spec.a"
+    // Example for struct: "a.b" to replace "desired.ko.Spec.a.b" with "<op_name>Latest.ko.spec.a.b"
+    // Example for map: "a.b.[c]" to replace "desired.ko.Spec.a.b[c]" with "<op_name>Latest.ko.spec.a.b[c]"
+    // TODO: support for List
+    
+    Path *string `json:"path"` 
+    
+    // TODO: possible support for source path and target path, which do not have to be exactly similar as they are at present.
+    }
```

Comment:
I was thinking that you would have a `LateInitializeConfig` struct that would store instructions for server-side defaulting (and eventually, server-side overrides). And that you would have a `late_initialize` field in the [`FieldConfig` struct](https://github.com/aws-controllers-k8s/code-generator/blob/c77aa9c75d944952dee198029ba9822691cd82b0/pkg/generate/config/field.go#L139-L187).

Something like this maybe...

```go
// LateInitializeConfig contains instructions for how to handle the retrieval and
// setting of server-side defaulted or overridden fields.
type LateInitializeConfig struct {
    // DefaultImmediately indicates that if the user has not specified a value for
    // the field, then directly after a successful call to Create, the reconciler will
    // call ReadOne to get the value of the field as defaulted by the AWS service,
    // set the value of this field on the resource and patch the resource back to
    // the Kubernetes API.
    DefaultImmediately bool `json:"default_immediately"`
}

...
type FieldConfig struct {
...
    LateInitialize *LateInitializeConfig `json:"late_initialize,omitempty"`
}
```

Reply:
I can move LateInitializationConfig to FieldConfig. 
My reasoning for placing it under resourceConfig was to capture all late-initialization at one place rather than spread around among multiple Field configuration. 
But I can also see why FieldConfig would make sense since it's a late initialization is property of the field.